provider/postgis/postgis.go

```diff
@@ -195,9 +203,144 @@ type hstoreOID struct {
 	hasInit bool
 }
 
+// validateURI validates for minimum requirements for a valid postgresql uri
+func validateURI(u string) error {
+	uri, err := url.Parse(u)
+	if err != nil {
+		return ErrInvalidURI{msg: fmt.Sprintf("%v", err)}
+	}
+
+	if uri.Scheme != "postgres" && uri.Scheme != "postgresql" {
+		return ErrInvalidURI{
+			msg: fmt.Sprintf("invalid connection scheme (%v)", uri.Scheme),
+		}
+	}
+
+	if uri.User == nil {
+		return ErrInvalidURI{msg: "auth credentials missing"}
+	}
+
+	host, port, err := net.SplitHostPort(uri.Host)
+	if err != nil {
+		return ErrInvalidURI{
+			msg: fmt.Sprintf("%v", err),
+		}
+	}
+
+	if host == "" {
+		return ErrInvalidURI{
+			msg: fmt.Sprintf("address %v:%v: missing host in address", host, port),
+		}
+	}
+
+	if uri.Path == "" {
+		return ErrInvalidURI{msg: "missing database"}
+	}
+
+	return nil
+}
+
+// BuildURI creates a database URI from config
+func BuildURI(config dict.Dicter) (*url.URL, *url.Values, error) {
+
+	sslmode := DefaultSSLMode
+	sslmode, err := config.String(ConfigKeySSLMode, &sslmode)
+	if err != nil {
+		return &url.URL{}, nil, err
+	}
+
+	uri := DefaultURI
+	uri, err = config.String(ConfigKeyURI, &uri)
+	if err != nil {
+		return &url.URL{}, nil, err
+	}
+
+	// if uri is set in the config, we add sslmode and return early
+	if uri != "" {
+		if err := validateURI(uri); err != nil {
+			return &url.URL{}, nil, err
```

Comment:
Why are you return a new url.URL object if there is an error; instead of just nil?